staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go

```diff
@@ -87,21 +87,21 @@ func (e *StatusError) DebugError() (string, []interface{}) {
 
 // HasStatusCause returns true if the provided error has a details cause
 // with the provided type name.
+// It supports wrapped errors and returns false when the error is nil.
 func HasStatusCause(err error, name metav1.CauseType) bool {
 	_, ok := StatusCause(err, name)
 	return ok
 }
 
 // StatusCause returns the named cause from the provided error if it exists and
-// the error is of the type APIStatus. Otherwise it returns false.
+// the error unwraps to the type APIStatus. Otherwise it returns false.
 func StatusCause(err error, name metav1.CauseType) (metav1.StatusCause, bool) {
-	apierr, ok := err.(APIStatus)
-	if !ok || apierr == nil || apierr.Status().Details == nil {
-		return metav1.StatusCause{}, false
-	}
-	for _, cause := range apierr.Status().Details.Causes {
-		if cause.Type == name {
-			return cause, true
+	var status APIStatus
+	if errors.As(err, &status) && status.Status().Details != nil {
```

Comment:
This is probably splitting hairs, but I was surprised to see that `errors.As` is implemented using reflection (notoriously slow). A quick benchmark confirmed that it's about an order of magnitude slower:
```
goos: linux
goarch: amd64
pkg: k8s.io/apimachinery/pkg/api/errors
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkTypeAssert_pos-8   	300704044	         3.950 ns/op
BenchmarkTypeAssert_neg-8   	277394091	         4.543 ns/op
BenchmarkErrorsAs_pos-8     	10509163	       118.8 ns/op
BenchmarkErrorsAs_neg-8     	35966838	        31.30 ns/op
```

I'm not sure if this is called in any hot paths, but I think it would be safest to just do the type assertion first, and then fallback to `errors.As`.

I'm honestly not sure how nitpicky this is, so feel free to weight for an apimachinery maintainer to weigh in.

Reply:
Understood. `errors.As` was introduced to other functions in this package during #88629. I don't see any discussion there about performance or why `StatusCause` might have been excluded.